### PR TITLE
Add support for PKCE and JWK-based authentication using "jose" library

### PR DESCRIPTION
**Pull request by jmandel**
Replacing proposal in https://github.com/smart-on-fhir/client-js/pull/139 with this branch that uses https://www.npmjs.com/package/jose to manage signatures. 

As a starting point:
```
$ du -h dist/build/fhir-client.min.js*
228K	dist/build/fhir-client.min.js
68K	dist/build/fhir-client.min.js.gz
```

### CODE REVIEW

**Comment by madaster97** [comment]
Hey Josh,

One question on this PR. Does the library stop you from providing/using a plain-text private key from the browser environment? I think adding that requirement/restriction would be valid so that it's more difficult for developers to expose static private keys on accident.

Thanks,
Adam

**Comment by jmandel** [comment]
Can you say more about what you mean by a plain text private key?  Are you thinking about making an assessment based on extractability? Or just thinking about disabling these methods from use on the browser altogether? We have never done that for symmetric authentication methods, and I'm not sure why the assistant here would change.

I could see value in logging a warning message in those cases though.

**Comment by madaster97** [comment]
One more comment. It looks like the PKCE support will fail if the server doesn't send "code_challenge_methods_supported" on it's well-known endpoint. Does that also mean it will fail if the server only supports the capability statement?

Put another way, will this PKCE support also work for SMART v1.0.0 servers, where that [discovery parameter](http://hl7.org/fhir/smart-app-launch/1.0.0/conformance/index.html#metadata) isn't defined?

**Comment by jmandel** [comment]
> Does that also mean it will fail if the server only supports the capability statement?

That is not compliant behavior in SMARTv1 or SMARTv2; servers need to support discovery via .well-known/smart-configuration

**Comment by madaster97** [comment]
Woops this wasn't what I meant to say:
> if the server only supports the capability statement

Putting that aside, [SMART v1.0.0](http://hl7.org/fhir/smart-app-launch/1.0.0/conformance/index.html#metadata) does not define the `code_challenge_methods_supported`. Is there any way this PR could also support that version?

**Comment by madaster97** [comment]
Woops this wasn't what I meant to say:
> if the server only supports the capability statement

Putting that aside, [SMART v1.0.0](http://hl7.org/fhir/smart-app-launch/1.0.0/conformance/index.html#metadata) does not define the `code_challenge_methods_supported`. Is there any way this PR could also support that version?

**Comment by jmandel** [comment]
You're welcome to pre-adopt properties and features from SMARTv2 in your SMARTv2 server - indeed, you can take them piecemeal if you like. Might be some better conversation for Zulip, as this is getting wide-rangind.

File: src/security/index.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+  cryptoRandomBytes = require('crypto').randomBytes
+} else {
+  wcrypto = window.crypto.subtle
+}
+
+export const digestSha256 = async (payload: string | ArrayBuffer) => {
+  let prepared: ArrayBuffer;
+
+  if (typeof payload === 'string') {
+    const encoder = new TextEncoder();
```
[the remark this answers is not in the document]
Comment (jmandel) [reply src/security/index.ts:26]: I think it's likely that EHRs will include PKCE support whenever they roll out SMARTv2, irrespective of whether their hospital customers are still using IE). I wonder if @madaster97 can comment?

[the remark this answers is not in the document]
Comment (madaster97) [reply src/security/index.ts:26]: Thanks for bringing this up. I tested this out and another pressing issue is that IE-11 doesn't support the current webcrypto standard. I'm trying out [isomorphic-webcrypto](https://www.npmjs.com/package/isomorphic-webcrypto) to see if it meets this need.

[the remark this answers is not in the document]
Comment (jmandel) [reply src/security/index.ts:26]: Good point. I think it's pretty unlikely that we will need confidential clients to authenticate from IE 11, but with PKC eterned on all clients will have no choice but to do *that*.

**Comment by vlad-ignatov** [comment]
I hope you saw my comment at https://github.com/microsoft-healthcare-madison/client-js/pull/3#issuecomment-1186549459 ?

Your changes have been merged into the [v3 branch](https://github.com/smart-on-fhir/client-js/tree/v3) for now, and then I made a lot of changes on top of them. However, the only significant difference is that the `security` file is now split in two - for browsers and for servers. `jose` is no longer in the browser because of its size, and because it can't be compiled for IE.

I'm going on a short vacation until 07/28 and will continue after that. Meanwhile, it would be helpful if you can check if the v3 branch works for you.

**Comment by madaster97** [comment]
Thanks @vlad-ignatov - I did see your comment but hadn't had a chance to look into things yet. I'll take a look at that and your changes this week

**Comment by madaster97** [comment]
Hi @vlad-ignatov ,

I'm actually stuck trying to use the v3 build. Mind taking a look?

I tried something similar to the docs [here](http://docs.smarthealthit.org/client-js/#browser-usage), but in edge it's not setting the "FHIR" variable. Any idea what I'm missing?:
```
<script src="https://raw.githubusercontent.com/smart-on-fhir/client-js/v3/dist/build/fhir-client.js"></script>
    <script>
      FHIR.oauth2.authorize({
        client_id: 'cc8d90f6-c7f8-4619-bd07-93301e892c04',
        scope: 'launch/patient patient/*.read',
        redirect_uri: 'http://localhost:5500/index.html',
        pkceMode: 'unsafeV1'
      })
    </script>
```

**Comment by jmandel** [comment]
Just to double check, does this example work for you in Chrome? (I have encountered CORS issues loading scripts directly from raw.githubusercontent.com.)

**Comment by madaster97** [comment]
It wasn't a CORS issue, but you were definitely right about that request being the issue. The response body (though a 200 status) is just `Unable to fetch script source`. I'll try again with a local copy

**Comment by madaster97** [comment]
Other than a comment I left [here](https://github.com/smart-on-fhir/client-js/commit/3ad1d83514ad24db4e6d6b16e7da08898b776a8f#r79701221), the v3 branch looks good to me 👍🏼 

I was able to get PKCE to work in IE-11 running in compatibility mode 11/edge. Mode 10 and below didn't work, as expected :
`<meta http-equiv="X-UA-Compatible" content="IE=edge">`

Thanks for the work here @vlad-ignatov 🥇 !

**Comment by vlad-ignatov** [comment]
You should be able to use URLs like https://combinatronics.com/smart-on-fhir/client-js/v3/dist/build/fhir-client.js

**Comment by vlad-ignatov** [comment]
This is now merged, then modified and released in `fhirclient@2.5.0`!

Despite having good test coverage, and even adding some tests in headless browser for Crypto stuff that cannot be mocked, I don't feel extremely confident about this since I have never tried in real life (in actual SMART app). I hope you can do that while I shift my focus to the launcher, and we can patch things as we go if needed.

**Comment by jmandel** [comment]
Just tested this out successfully as a drop-in replacement for the client in a remixed version fo https://glitch.com/edit/#!/sharp-lake-word -- nice! A couple of notes:

* For https://github.com/smart-on-fhir/client-js/blob/master/src/security/browser.ts#L88 is there a reason not to use `new TextEncoder().encode(` instead of these custom functions? If this is an "IE11" issue, it's one where I'd rather streamline and just say "asymmetric signatures aren't supported in the SMART JS client for IE11".

* If we're going to include a console warning at https://github.com/smart-on-fhir/client-js/blob/5d3ab96e5100b7b3aa5e45d9fe032f051e52c2f7/src/smart.ts#L677-L690 it should fire when private JWK was loaded 4 lines earlier, since that too could be stealable even if the imported key object isn't "extractable", and the fix is the same (to provision and register non-extractable keys).

**Comment by vlad-ignatov** [comment]
- Regarding `s2b` (which stands for string to buffer) - yes, that was an IE11 thing. The assumption was that this should work in IE11. I am OK with simplifying it and dropping asymmetric auth from IE, but the problem is that `s2b` is also needed for generating codeChallenge for PKCE.
- As for the stealable warning, just to be clear, are you saying that importing of any JWK should trigger that warning?

**Comment by jmandel** [comment]
>  s2b is also needed for generating codeChallenge for PKCE.

Good point. I still think it's OK to cut IE11 support for these "new SMARTv2 features" though. Like, we should continue to support folks already using the library for an older app, but IE11 has reached end of life and I'd rather not take on new technical debt to support it.

> As for the stealable warning, just to be clear, are you saying that importing of any JWK should trigger that warning?

If we're calling `env.security.importJWK(privateKey as fhirclient.JWK)` *as part of the token request*, this implies the key is available in memory for client code to read, so it's the same risk surface as an extractable `CryptoKey` object. At least, that seems like the case to me. Do you agree?

**Comment by vlad-ignatov** [comment]
> cut IE11 support for these "new SMARTv2 features"

Do you also mean no PKCE in IE?

> this implies the key is available in memory for client code to read

I agree. This is the part that confuses me a little though. We allow people to provide JWK and we use it as is. In other words, we end up allowing insecure features, and on top of that we show warnings that cannot be avoided unless they stop using those features!?

**Comment by jmandel** [comment]
>In other words, we end up allowing insecure features

I wouldn't call this insecure, and I'm not sure we need a warning at all. It is strictly more secure than the confidential client profiles widely used today, because it allows for authentication without having to send a secret over the wire. Of course it would be unusual to deploy a confidential client that runs purely in the browser, unless the EHR supports some kind of instance level client registration. But this is a separate point.

Bottom line: this concern about credential security only applies if you are running untrusted code in your app's origin ... and if you're doing that, you are already in trouble because you are exposing clinical data etc. If we are going to include a warning it should cover all cases that expose the user to this risk. But based on this discussion, I would just omit the the runtime notification and instead provide developer documentation about these considerations.

>Do you also mean no PKCE in IE?

Yes that would be the implication. "No support for SMART V2 in IE 11" seems like a reasonable place to make a clean break.

**Comment by vlad-ignatov** [comment]
OK. All that, plus an important related bug-fix (#158) is now released in `fhirclient@2.5.2`

**Pull request merged**